bpo-42061: Document __format__ for IP addresses

John-Ted opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-42061](https://bugs.python.org/issue42061) -->
https://bugs.python.org/issue42061
<!-- /issue-number -->

Automerge-Triggered-By: GH:ericvsmith

John-Ted committed:
Document __format__ for IP addresses

Fix issue #42061

John-Ted committed:
Add false positive to tools\susp-ignored.csv

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

John-Ted committed:
Remove unnecessary rules in tests/susp-ignored.csv

ericvsmith commented on Doc/library/ipaddress.rst:
```diff
@@ -317,6 +317,32 @@ the :func:`str` and :func:`int` builtin functions::
 Note that IPv6 scoped addresses are converted to integers without scope zone ID.
 
 
+Other formatting options
```
I think we should add `__format__` under the list of methods and document this there, similar to how datetime does it. As written, this section doesn't mention `__format__`. Unless you know that `format()` calls `__format__` (and most people don't), then I'm not sure you'd make the connection that these format options are also available with `str.format()` and f-strings.

John-Ted committed:
Implement reviewer suggestions

ericvsmith commented on Doc/library/ipaddress.rst:
```diff
@@ -199,9 +199,34 @@ write code that handles both IP versions correctly.  Address objects are
+      representations, the form specifier ``'#'`` and the grouping option
+      ``'_'`` are available.
+
+         >>> ipaddress.IPv4Address('192.168.0.1').__format__('s')
```
Having said that I think the documentation belongs under `__format__` (which I still think is the right thing to do), I'm not sure using `__format__` in the examples is the right thing. We discourage people from calling dunders directly. Maybe there should be a note that `__format__` is used by `format`, `str.format`, and f-strings, and then show examples using one or more of those? I'm open to suggestions here.

John-Ted committed:
Fix formatting

ericvsmith commented on Doc/library/ipaddress.rst:
```diff
@@ -199,9 +199,34 @@ write code that handles both IP versions correctly.  Address objects are
 .. _iana-ipv4-special-registry: https://www.iana.org/assignments/iana-ipv4-special-registry/iana-ipv4-special-registry.xhtml
 .. _iana-ipv6-special-registry: https://www.iana.org/assignments/iana-ipv6-special-registry/iana-ipv6-special-registry.xhtml
-
+   
```
Extraneous whitespace here.

John-Ted committed:
Implement reviewer suggestions

The examples no longer use the dunder directly and the
__format__ documentation is changed to be like that of datetime
(as suggested in the first review)

John-Ted committed:
Change ipaddress to IPv[46]Address in method doc

ericvsmith reviewed (approved):
Looks great to me! Thanks for your patience.

ericvsmith commented on Doc/library/ipaddress.rst:
```diff
@@ -202,6 +202,33 @@ write code that handles both IP versions correctly.  Address objects are
+      >>> '{:#b}'.format(ipaddress.IPv4Address('192.168.0.1'))
+      '0b11000000101010000000000000000001'
+      >>> ipv6_address = ipaddress.IPv6Address('2001:db8::1000')
+      >>> f'{ipv6_address:'s'}'
```
Oops, I spoke too soon. There's a typo in the f-string that the automated doctest bot is objecting to.

John-Ted committed:
Fix bug in fstring example

John-Ted committed:
Add false positives to susp-ignored.csv
